library/Freckle/App/Http/Cache.hs

```diff
@@ -0,0 +1,304 @@
+import Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
+import Freckle.App.Http.Header
+import Freckle.App.Memcached
+import Network.HTTP.Client (Request, Response)
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Simple
+  ( addRequestHeader
+  , getRequestHeader
+  , getResponseStatus
+  )
+import Network.HTTP.Types.Header
+  ( HeaderName
+  , hAge
+  , hCacheControl
+  , hETag
+  , hExpires
+  , hIfNoneMatch
+  , hVary
+  )
+import Network.HTTP.Types.Status (Status, statusCode)
+
+data HttpCacheSettings m body t = HttpCacheSettings
+  { shared :: Bool
+  , cacheable :: Request -> Bool
+  , defaultTTL :: CacheTTL
+  , getCurrentTime :: m UTCTime
+  , logWarn :: String -> m ()
+  , codec :: HttpCacheCodec body t
+  , cache :: HttpCache m t
+  }
+
+data HttpCacheCodec body t = HttpCacheCodec
+  { serialise :: CachedResponse body -> t
+  , deserialise :: Request -> t -> Either String (CachedResponse body)
+  }
+
+data HttpCache m t = HttpCache
+  { get :: CacheKey -> m (Either SomeException (Maybe t))
+  , set :: CacheKey -> t -> m (Either SomeException ())
+  , evict :: CacheKey -> m (Either SomeException ())
+  }
+
+data CachedResponse body = CachedResponse
+  { response :: Response body
+  , inserted :: UTCTime
+  , ttl :: CacheTTL
+  }
+  deriving stock (Show)
+
+isCachedResponseStale :: CachedResponse body -> UTCTime -> Bool
+isCachedResponseStale cached now =
+  addUTCTime (fromIntegral cached.ttl) cached.inserted < now
+
+-- Wrap a function from "Freckle.App.Http" with caching
+--
+-- Verify that the request is cacheable (e.g. a @GET@), then cache it at a
+-- derived key (from URL and considering any @Vary@ headers). The response will
+-- only be cached if @Cache-Control@ allows it. @Cache-Control@ is also used to
+-- determine TTL (e.g. @max-age@)
+--
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#vary>
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#fresh_and_stale_based_on_age>
+--
+-- If a cached response is stale, but it has an @ETag@ header, we will make the
+-- request using @If-None-Match@ and still return (and retain) that cached
+-- response if we receive a @304@ response.
+--
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#caching_of_unchanged_resources>
+--
+-- Usage:
+--
+-- @
+-- import "Freckle.App.Http"
+-- import "Freckle.App.Http.Cache"
+-- import "Freckle.App.Http.Cache.Memcached"
+--
+-- data Foo = Foo
+--   { -- ...
+--   }
+--   deriving stock Generic
+--   deriving anyclass (Serialise, ToJSON)
+--
+-- getFooById
+--   :: ( 'MonadUnliftIO' m
+--      , 'MonadHttp' m
+--      , 'MonadLogger' m
+--      , 'MonadTracer' m
+--      , 'MonadReader' env m
+--      , 'HasMemcachedClient' env
+--      )
+--   => FooId -> m [Foo]
+-- getFooById fooId = do
+--   let req = parseRequest_ $ \"https:\/\/api.com\/foos\/\" <> fooId
+--   'httpCached' ('memcachedHttpCacheSettings' 'fiveMinuteTTL') 'httpJson' req
```

Comment:
I don't _actually_ know if this works yet. It would need `Serialise (Either DecodeError a)`, but if that instance could exist, then passing `httpJson` as the `Request -> m (Response body)` like this would totally work.